<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>DRAM2 User Manual &mdash; DRAM2 0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=2709fde1"></script>
        <script src="_static/doctools.js?v=888ff710"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Nextflow Pipeline Options" href="nextflow_options.html" />
    <link rel="prev" title="Output Files" href="example_output.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #1e4b28" >

          
          
          <a href="index.html">
            
              <img src="_static/dram_logo_med_white.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quick_start.html">Quick Start Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="example_output.html">Output Files</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">DRAM2 User Manual</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#bring-up-help-menu-s">Bring up help menu(s)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#bring-up-versions-menu">Bring up versions menu</a></li>
<li class="toctree-l2"><a class="reference internal" href="#input-data">Input data</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#starting-from-scratch-provide-dram2-with-input-fasta-files">Starting from scratch: Provide DRAM2 with input FASTA files</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#fasta-files">FASTA files:</a></li>
<li class="toctree-l4"><a class="reference internal" href="#optional-additional-files">(Optional) Additional files:</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sample-taxonomy">Sample taxonomy:</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sample-bin-mag-quality">Sample Bin/MAG quality:</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#starting-with-annotation-provide-dram2-with-input-called-genes">Starting with Annotation: Provide DRAM2 with input called genes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#starting-with-distill-provide-dram2-with-previous-annotations">Starting with Distill: Provide DRAM2 with previous annotations</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#general-command-line-options">General Command-line Options</a></li>
<li class="toctree-l2"><a class="reference internal" href="#call-command-line-options">Call Command-line Options</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#call-description">Call description</a></li>
<li class="toctree-l3"><a class="reference internal" href="#usage">Usage</a></li>
<li class="toctree-l3"><a class="reference internal" href="#required-dram2-profile-options">REQUIRED DRAM2 profile options</a></li>
<li class="toctree-l3"><a class="reference internal" href="#call-options">Call options</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#annotate-command-line-options">Annotate Command-line Options</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#annotate-description">Annotate description</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id1">Usage</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id2">REQUIRED DRAM2 profile options</a></li>
<li class="toctree-l3"><a class="reference internal" href="#annotate-options">Annotate options</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#distill-command-line-options">Distill Command-line Options</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#distill-description">Distill description</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id3">Usage</a></li>
<li class="toctree-l3"><a class="reference internal" href="#example">Example</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id4">REQUIRED DRAM2 profile options</a></li>
<li class="toctree-l3"><a class="reference internal" href="#distill-options">Distill options</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#select-a-profile-which-suits-your-compute-setup">Select a Profile which suits your compute setup</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#which-is-better">Which is Better?</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#conda-environments">Conda Environments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#singularity-containers">Singularity Containers</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#dram2-databases">DRAM2 Databases</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#all-databases-dram2-accommodates">All databases DRAM2 accommodates:</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#downloadable-dram2-database-sets">Downloadable DRAM2 Database Sets</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#slurm-options">SLURM Options</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#slurm-command-line-options">SLURM command-line options</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#software-used">Software Used</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="nextflow_options.html">Nextflow Pipeline Options</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #1e4b28" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">DRAM2</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">DRAM2 User Manual</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/user_manual.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="dram2-user-manual">
<h1>DRAM2 User Manual<a class="headerlink" href="#dram2-user-manual" title="Link to this heading">ÔÉÅ</a></h1>
<p>DRAM2 provides various commands for calling genes, annotating called genes, and distilling annotations from input FASTA files. Below are the available commands:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nextflow<span class="w"> </span>run<span class="w"> </span>DRAM2.nf<span class="w"> </span>--help<span class="w"> </span><span class="o">(</span>--h<span class="o">)</span>

nextflow<span class="w"> </span>run<span class="w"> </span>DRAM2.nf<span class="w"> </span>--version<span class="w"> </span><span class="o">(</span>--v<span class="o">)</span>

Example:

nextflow<span class="w"> </span>run<span class="w"> </span>DRAM2.nf<span class="w"> </span>--input_fasta<span class="w"> </span>../test_data/
<span class="w">    </span>--outdir<span class="w"> </span>DRAM2-test-data-Feb012024/
<span class="w">    </span>--call<span class="w"> </span>--rename
<span class="w">    </span>--annotate<span class="w"> </span>--use_uniref<span class="w"> </span>--use_kegg<span class="w"> </span>--use_merops<span class="w"> </span>--use_viral<span class="w"> </span>--use_camper<span class="w"> </span>--use_kofam<span class="w"> </span>--use_dbcan<span class="w"> </span>--use_methyl<span class="w"> </span>--use_canthyd<span class="w"> </span>--use_vog<span class="w"> </span>--use_fegenie<span class="w"> </span>--use_sulfur
<span class="w">    </span>--add_annotations<span class="w"> </span>../test-data/old-DRAM-annotations.tsv
<span class="w">    </span>--distill_topic<span class="w"> </span><span class="s1">&#39;carbon transport energy&#39;</span><span class="w"> </span>--distill_ecosystem<span class="w"> </span><span class="s1">&#39;eng_sys ag&#39;</span>
<span class="w">    </span>--distill_custom<span class="w"> </span>assets/forms/distill_sheets/test.tsv<span class="w"> </span>-resume<span class="w"> </span>--slurm_node<span class="w"> </span>zenith
<span class="w">    </span>--trnas<span class="w"> </span>../test-data/trnas.tsv
<span class="w">    </span>--rrnas<span class="w"> </span>../test-data/rrnas.tsv
<span class="w">    </span>--bin_quality<span class="w"> </span>../test-data/checkM1-test-data.tsv
<span class="w">    </span>--taxa<span class="w"> </span>../test-data/gtdbtk.bac120.summary.tsv
<span class="w">    </span>--generate_gff
<span class="w">    </span>--generate_gbk
<span class="w">    </span>--threads<span class="w"> </span><span class="m">5</span>
<span class="w">    </span>-with-report<span class="w"> </span>-with-trace<span class="w"> </span>-with-timeline
</pre></div>
</div>
<section id="bring-up-help-menu-s">
<h2>Bring up help menu(s)<a class="headerlink" href="#bring-up-help-menu-s" title="Link to this heading">ÔÉÅ</a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nextflow<span class="w"> </span>run<span class="w"> </span>DRAM2.nf<span class="w"> </span>--help<span class="w"> </span><span class="o">(</span>--h<span class="o">)</span>
</pre></div>
</div>
</section>
<section id="bring-up-versions-menu">
<h2>Bring up versions menu<a class="headerlink" href="#bring-up-versions-menu" title="Link to this heading">ÔÉÅ</a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nextflow<span class="w"> </span>run<span class="w"> </span>DRAM2.nf<span class="w"> </span>--version<span class="w"> </span><span class="o">(</span>--v<span class="o">)</span>
</pre></div>
</div>
</section>
<hr class="docutils" />
<section id="input-data">
<h2>Input data<a class="headerlink" href="#input-data" title="Link to this heading">ÔÉÅ</a></h2>
<section id="starting-from-scratch-provide-dram2-with-input-fasta-files">
<h3>Starting from scratch: Provide DRAM2 with input FASTA files<a class="headerlink" href="#starting-from-scratch-provide-dram2-with-input-fasta-files" title="Link to this heading">ÔÉÅ</a></h3>
<section id="fasta-files">
<h4>FASTA files:<a class="headerlink" href="#fasta-files" title="Link to this heading">ÔÉÅ</a></h4>
<ul class="simple">
<li><p>FASTA files must be located in a single directory</p></li>
<li><p>A given FASTA file represents one sample/MAG/isolate</p></li>
<li><p>FASTA files have one of the following extensions: .fa, .fasta, .fna</p></li>
<li><p>FASTA headers are either renamed with respect to the individual sample or you will plan to use the <cite>‚Äìrename</cite> option</p></li>
</ul>
</section>
<section id="optional-additional-files">
<h4>(Optional) Additional files:<a class="headerlink" href="#optional-additional-files" title="Link to this heading">ÔÉÅ</a></h4>
</section>
<section id="sample-taxonomy">
<h4>Sample taxonomy:<a class="headerlink" href="#sample-taxonomy" title="Link to this heading">ÔÉÅ</a></h4>
<ul class="simple">
<li><p>Sample taxonomy input (<code class="docutils literal notranslate"><span class="pre">--taxa</span> <span class="pre">&lt;path/to/file&gt;</span></code>) is modeled after <a class="reference external" href="https://ecogenomics.github.io/GTDBTk/files/summary.tsv.html#files-summary-tsv">GTDB-Tk output</a></p></li>
<li><p>Minimum requirements: 2 columns named, ‚Äúuser_genome‚Äù (sample names matching input file names) and ‚Äúclassification‚Äù (taxonomy)</p></li>
<li><p>Taxonomy information will be added to the output <cite>raw-annotations.tsv</cite> and <cite>distillate.xlsx</cite></p></li>
</ul>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>user_genome</p></th>
<th class="head"><p>classification</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>sample1</p></td>
<td><p>taxonomy1</p></td>
</tr>
<tr class="row-odd"><td><p>sample2</p></td>
<td><p>taxonomy2</p></td>
</tr>
<tr class="row-even"><td><p>sample3</p></td>
<td><p>taxonomy3</p></td>
</tr>
</tbody>
</table>
<p><strong>Note:</strong> DRAM2 visualization may rely on GTDB-Tk-formatted taxonomy and it is in your best interest to follow this format.</p>
</section>
<section id="sample-bin-mag-quality">
<h4>Sample Bin/MAG quality:<a class="headerlink" href="#sample-bin-mag-quality" title="Link to this heading">ÔÉÅ</a></h4>
<ul class="simple">
<li><p>Sample MAG/Bin quality input (<code class="docutils literal notranslate"><span class="pre">--bin_quality</span> <span class="pre">&lt;path/to/file&gt;</span></code>) is modeled after <a class="reference external" href="https://github.com/chklovski/CheckM2">CheckM/CheckM2 output</a></p></li>
<li><p>Minimum requirements: 3 columns named, ‚ÄúBin Id‚Äù (sample names matching input file names), ‚ÄúCompleteness‚Äù and ‚ÄúContamination‚Äù</p></li>
<li><p>Bin quality information will be added to the output <cite>raw-annotations.tsv</cite> and <cite>distillate.xlsx</cite></p></li>
</ul>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Bin Id</p></th>
<th class="head"><p>Completeness</p></th>
<th class="head"><p>Contamination</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>sample1</p></td>
<td><p>95%</p></td>
<td><p>2%</p></td>
</tr>
<tr class="row-odd"><td><p>sample2</p></td>
<td><p>88%</p></td>
<td><p>5%</p></td>
</tr>
<tr class="row-even"><td><p>sample3</p></td>
<td><p>92%</p></td>
<td><p>1.5%</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="starting-with-annotation-provide-dram2-with-input-called-genes">
<h3>Starting with Annotation: Provide DRAM2 with input called genes<a class="headerlink" href="#starting-with-annotation-provide-dram2-with-input-called-genes" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Input called genes (proteins), ideally from the latest version of Prodigal</p></li>
<li><p>Input called genes must have the extension: .faa</p></li>
<li><p>Input called gene headers contain the sample-name prefix (this is crucial to ensure annotations are sample-specific)</p></li>
</ul>
<p>Example:
Sample01.faa has the following First entry:</p>
<blockquote>
<div><div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>&gt;Sample01_k95_149252_1<span class="w"> </span><span class="c1"># 1 # 513 # 1 # ID=1_1;partial=10;start_type=Edge;rbs_motif=None;rbs_spacer=None;gc_cont=0.392</span>
VVNQLANNRQGTQSAKTRSEVSGGGRKPWRQKGTGHARQGSTRSPQWTGGGVVFAPKPRD
</pre></div>
</div>
</div></blockquote>
<p><strong>Note:</strong> DRAM2 uses rename.sh from <a class="reference external" href="https://jgi.doe.gov/data-and-tools/bbtools/">BBTools</a></p>
</section>
<section id="starting-with-distill-provide-dram2-with-previous-annotations">
<h3>Starting with Distill: Provide DRAM2 with previous annotations<a class="headerlink" href="#starting-with-distill-provide-dram2-with-previous-annotations" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Input annotations (TSV) file must be generated via DRAM2</p></li>
<li><p>(Optional) rRNAs (<code class="docutils literal notranslate"><span class="pre">--rrnas</span> <span class="pre">&lt;path/to/file&gt;</span></code>) identified with <a class="reference external" href="https://github.com/tseemann/barrnap">Barrnap</a></p></li>
<li><p>(Optional) tRNAs  (<code class="docutils literal notranslate"><span class="pre">--trnas</span> <span class="pre">&lt;path/to/file&gt;</span></code>) identified with <a class="reference external" href="http://lowelab.ucsc.edu/tRNAscan-SE/">tRNAscan-SE</a></p></li>
<li><p>tRNA and rRNA information is included in the output <cite>distillate.xlsx</cite></p></li>
</ul>
</section>
</section>
<hr class="docutils" />
<section id="general-command-line-options">
<h2>General Command-line Options<a class="headerlink" href="#general-command-line-options" title="Link to this heading">ÔÉÅ</a></h2>
<p>The usage of DRAM2 involves various commands and options, as outlined below:</p>
<ul>
<li><p>To perform a combined operation of renaming, calling, annotating, and distilling, use:</p>
<blockquote>
<div><div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nextflow<span class="w"> </span>run<span class="w"> </span>DRAM2.nf<span class="w"> </span>--rename<span class="w"> </span>--call<span class="w"> </span>--annotate<span class="w"> </span>--use_&lt;database<span class="o">(</span>s<span class="o">)</span>&gt;<span class="w"> </span>--distill_topic<span class="w"> </span>&lt;distillate<span class="o">(</span>s<span class="o">)</span>&gt;
</pre></div>
</div>
</div></blockquote>
</li>
<li><p>To call genes using input FASTA files and optionally rename the FASTA headers, use:</p>
<blockquote>
<div><div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nextflow<span class="w"> </span>run<span class="w"> </span>DRAM2.nf<span class="w"> </span>--call<span class="w"> </span>--rename<span class="w"> </span>--input_fasta_dir<span class="w"> </span>&lt;path/to/fasta/directory/&gt;
</pre></div>
</div>
</div></blockquote>
</li>
<li><p>To annotate called genes using input FASTA files, use:</p>
<blockquote>
<div><div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nextflow<span class="w"> </span>run<span class="w"> </span>DRAM2.nf<span class="w"> </span>--annotate<span class="w"> </span>--input_genes<span class="w"> </span>&lt;path/to/called/genes/directory&gt;
</pre></div>
</div>
</div></blockquote>
</li>
<li><p>To distill annotations using input annotation files, use:</p>
<blockquote>
<div><div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nextflow<span class="w"> </span>run<span class="w"> </span>DRAM2.nf<span class="w"> </span>--distill_&lt;topic<span class="p">|</span>ecosystem<span class="p">|</span>custom&gt;<span class="w"> </span>--annotations<span class="w"> </span>&lt;path/to/annotations.tsv&gt;
</pre></div>
</div>
</div></blockquote>
</li>
<li><p>For a more realistic example with multiple options, see:</p>
<blockquote>
<div><div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nextflow<span class="w"> </span>run<span class="w"> </span>DRAM2.nf
<span class="w">    </span>--input_fasta<span class="w"> </span>../test_data/
<span class="w">    </span>--outdir<span class="w"> </span>DRAM2-test-data-Feb012024/
<span class="w">    </span>--call<span class="w"> </span>--rename
<span class="w">    </span>--annotate<span class="w"> </span>--use_uniref<span class="w"> </span>--use_kegg<span class="w"> </span>--use_merops<span class="w"> </span>--use_viral<span class="w"> </span>--use_camper<span class="w"> </span>--use_kofam<span class="w"> </span>--use_dbcan<span class="w"> </span>--use_methyl<span class="w"> </span>--use_canthyd<span class="w"> </span>--use_vog<span class="w"> </span>--use_fegenie<span class="w"> </span>--use_sulfur
<span class="w">    </span>--add_annotations<span class="w"> </span>../test-data/old-DRAM2-annotations.tsv
<span class="w">    </span>--distill_topic<span class="w"> </span><span class="s1">&#39;carbon transport energy&#39;</span><span class="w"> </span>--distill_ecosystem<span class="w"> </span><span class="s1">&#39;eng_sys ag&#39;</span>
<span class="w">    </span>--distill_custom<span class="w"> </span>assets/forms/distill_sheets/test.tsv<span class="w"> </span>-resume<span class="w"> </span>--slurm_node<span class="w"> </span>zenith
<span class="w">    </span>--trnas<span class="w"> </span>../test-data/trnas.tsv
<span class="w">    </span>--rrnas<span class="w"> </span>../test-data/rrnas.tsv
<span class="w">    </span>--bin_quality<span class="w"> </span>../test-data/checkM1-test-data.tsv
<span class="w">    </span>--taxa<span class="w"> </span>../test-data/gtdbtk.bac120.summary.tsv
<span class="w">    </span>--generate_gff
<span class="w">    </span>--generate_gbk
<span class="w">    </span>--threads<span class="w"> </span><span class="m">5</span>
<span class="w">    </span>-with-report<span class="w"> </span>-with-trace<span class="w"> </span>-with-timeline
<span class="w">    </span>-profile<span class="w"> </span>conda_slurm
<span class="w">    </span>-slurm_node<span class="w"> </span>alpha
</pre></div>
</div>
</div></blockquote>
</li>
</ul>
</section>
<hr class="docutils" />
<section id="call-command-line-options">
<h2>Call Command-line Options<a class="headerlink" href="#call-command-line-options" title="Link to this heading">ÔÉÅ</a></h2>
<section id="call-description">
<h3>Call description<a class="headerlink" href="#call-description" title="Link to this heading">ÔÉÅ</a></h3>
<p>The purpose of DRAM2 ‚Äìcall is to call genes on input FASTA files.</p>
</section>
<section id="usage">
<h3>Usage<a class="headerlink" href="#usage" title="Link to this heading">ÔÉÅ</a></h3>
<p>To call genes using input FASTA files, use the following command:
.. code-block:: bash</p>
<blockquote>
<div><p>nextflow run DRAM2.nf ‚Äìcall ‚Äìinput_fasta_dir &lt;path/to/fasta/directory/&gt; ‚Äìoutdir &lt;path/to/output/directory/&gt; ‚Äìthreads &lt;threads&gt;</p>
</div></blockquote>
</section>
<section id="required-dram2-profile-options">
<h3>REQUIRED DRAM2 profile options<a class="headerlink" href="#required-dram2-profile-options" title="Link to this heading">ÔÉÅ</a></h3>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">-profile</span></code>: STRING &lt;conda, conda_slurm, singularity, singularity_conda&gt;</p>
<blockquote>
<div><p>Runs DRAM2 either using Conda (must be installed) or Singularity (must be installed).
Runs DRAM2 with no scheduling or scheduling via SLURM.
See SLURM options in the full help menu.</p>
</div></blockquote>
</li>
</ul>
</section>
<section id="call-options">
<h3>Call options<a class="headerlink" href="#call-options" title="Link to this heading">ÔÉÅ</a></h3>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">--rename</span></code>: Rename FASTA headers based on file name.</p>
<blockquote>
<div><p>Why? DRAM2 output is focused on scaffolds/contigs with respect to each provided input sample.
Thus, without renaming FASTA headers, the individual scaffolds/contigs will not be distinguishable.
<em>If you have already renamed your FASTA headers, do not include ‚Äò‚Äìcall‚Äô</em></p>
</div></blockquote>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">--prodigal_mode</span></code>: STRING &lt;single|meta&gt;</p>
<blockquote>
<div><p>Default: ‚Äòsingle‚Äô</p>
</div></blockquote>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">--prodigal_tras_table</span></code>: &lt;1|2|3|4|5|6|7|8|9|10|11|12|13|14|15|16|17|18|19|20|21|22|23|24|25&gt;</p>
<blockquote>
<div><p>Specify a translation table to use (default: ‚Äò1‚Äô).</p>
</div></blockquote>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">--min_contig_len</span></code>: NUMBER &lt;number in base pairs&gt;</p>
<blockquote>
<div><p>Default: ‚Äò2500‚Äô</p>
</div></blockquote>
</li>
</ul>
</section>
</section>
<hr class="docutils" />
<section id="annotate-command-line-options">
<h2>Annotate Command-line Options<a class="headerlink" href="#annotate-command-line-options" title="Link to this heading">ÔÉÅ</a></h2>
<section id="annotate-description">
<h3>Annotate description<a class="headerlink" href="#annotate-description" title="Link to this heading">ÔÉÅ</a></h3>
<p>The purpose of DRAM2 ‚Äò‚Äìannotate‚Äô is to annotate called genes on input (nucleotide) FASTA (fa*) files.</p>
</section>
<section id="id1">
<h3>Usage<a class="headerlink" href="#id1" title="Link to this heading">ÔÉÅ</a></h3>
<p>To annotate called genes, use one of the following commands:
.. code-block:: bash</p>
<blockquote>
<div><p>Annotate called genes using input called genes and the KOFAM database:
nextflow run DRAM2.nf ‚Äìannotate ‚Äìinput_genes &lt;path/to/called/genes/directory&gt; ‚Äìuse_kofam</p>
<p>Annotate called genes using input fasta files and the KOFAM database:
nextflow run DRAM2.nf ‚Äìannotate ‚Äìinput_fasta &lt;path/to/called/genes/directory&gt; ‚Äìuse_kofam</p>
</div></blockquote>
</section>
<section id="id2">
<h3>REQUIRED DRAM2 profile options<a class="headerlink" href="#id2" title="Link to this heading">ÔÉÅ</a></h3>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">-profile</span></code>: <cite>&lt;conda, conda_slurm, singularity, singularity_conda&gt;</cite></p>
<blockquote>
<div><p>Runs DRAM2 either using Conda (must be installed) or Singularity (must be installed).
Runs DRAM2 with no scheduling or scheduling via SLURM.
<em>See SLURM options in the full help menu.</em></p>
</div></blockquote>
</li>
</ul>
</section>
<section id="annotate-options">
<h3>Annotate options<a class="headerlink" href="#annotate-options" title="Link to this heading">ÔÉÅ</a></h3>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">--use_&lt;db-name&gt;</span></code>: STRING &lt;camper|cant_hyd|dbcan|fegenie|kegg|kofam|merops|methyl|heme|pfam|sulfur|uniref&gt;</p>
<blockquote>
<div><p>Specify databases to use. Can use more than one. Can be used in combination with ‚Äìuse_dbset.</p>
</div></blockquote>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">--use_dbset</span></code>: STRING &lt;metabolism_kegg_set|metabolism_set|adjectives_kegg_set|adjectives_set&gt;</p>
<blockquote>
<div><p>metabolism_kegg_set = kegg, dbcan, merops, pfam, heme
metabolism_set        = kofam, dbcan, merops, pfam, heme
adjectives_kegg_set = kegg, dbcan, merops, pfam, heme, sulfur, camper, methyl, fegenie
adjectives_set        = kofam, dbcan, merops, pfam, heme, sulfur, camper, methyl, fegenie
<em>Only one set can be used. Can be used in combination with ‚Äìuse_[db-name]</em></p>
</div></blockquote>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">--add_annotations</span></code>: PATH &lt;path/to/old-annoations.tsv&gt;</p>
<blockquote>
<div><p>Used to add in old annotations to the current run. (See example for format.)</p>
</div></blockquote>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">--generate_gff</span></code>: OPTION</p>
<blockquote>
<div><p>Will generate an output GFF for each sample based on the raw-annotations.tsv.</p>
</div></blockquote>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">--generate_gbk</span></code>: OPTION</p>
<blockquote>
<div><p>Will generate an output GBK for each sample based on the raw-annotations.tsv.</p>
</div></blockquote>
</li>
</ul>
</section>
</section>
<hr class="docutils" />
<section id="distill-command-line-options">
<h2>Distill Command-line Options<a class="headerlink" href="#distill-command-line-options" title="Link to this heading">ÔÉÅ</a></h2>
<section id="distill-description">
<h3>Distill description<a class="headerlink" href="#distill-description" title="Link to this heading">ÔÉÅ</a></h3>
<p>The purpose of DRAM2 ‚Äìdistill is to distill down annotations based on curated distillation summary form(s).
User‚Äôs may also provide a custom distillate via ‚Äìdistill_custom &lt;path/to/file&gt; (TSV forms).
Distill can be run independently of ‚Äìcall and ‚Äìannotate; however, annotations must be provided (‚Äìannotations &lt;path/to/annotations.tsv&gt;).
Optional tRNA, rRNA, and bin quality may also be provided.</p>
</section>
<section id="id3">
<h3>Usage<a class="headerlink" href="#id3" title="Link to this heading">ÔÉÅ</a></h3>
<p>To distill annotations, use the following command:
.. code-block:: bash</p>
<blockquote>
<div><p>nextflow run DRAM2.nf ‚Äìdistill_&lt;topic|ecosystem|custom&gt; ‚Äìannotations &lt;path/to/annotations.tsv&gt; ‚Äìoutdir &lt;path/to/output/directory/&gt; ‚Äìthreads &lt;threads&gt;</p>
</div></blockquote>
<p><em>Important: if more than one topic or ecosystem is included, they must be enclosed in single quotes. Example: ‚Äìdistill_topic ‚Äòcarbon transport‚Äô</em></p>
</section>
<section id="example">
<h3>Example<a class="headerlink" href="#example" title="Link to this heading">ÔÉÅ</a></h3>
<p>Call and Annotate genes using input fastas and KOFAM database. Distill using carbon topic and AG ecosystem:
.. code-block:: bash</p>
<blockquote>
<div><p>nextflow run DRAM2.nf ‚Äìinput_fasta_dir &lt;path/to/fasta/directory/&gt; ‚Äìoutdir &lt;path/to/output/directory/&gt; ‚Äìcall ‚Äìannotate ‚Äìdistill_topic carbon ‚Äìdistill_ecosystem ag ‚Äìthreads &lt;threads&gt; ‚Äìuse_kofam</p>
</div></blockquote>
</section>
<section id="id4">
<h3>REQUIRED DRAM2 profile options<a class="headerlink" href="#id4" title="Link to this heading">ÔÉÅ</a></h3>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">-profile</span></code>: STRING &lt;conda, conda_slurm, singularity, singularity_conda&gt;</p>
<blockquote>
<div><p>Runs DRAM2 either using Conda (must be installed) or Singularity (must be installed).
Runs DRAM2 with no scheduling or scheduling via <a class="reference external" href="https://slurm.schedmd.com/documentation.html">SLURM</a>.
See SLURM options in the full help menu.</p>
</div></blockquote>
</li>
</ul>
</section>
<section id="distill-options">
<h3>Distill options<a class="headerlink" href="#distill-options" title="Link to this heading">ÔÉÅ</a></h3>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">--annotations</span></code>: PATH &lt;path/to/annotations.tsv&gt;</p>
<blockquote>
<div><p>Required if you are running distill without ‚Äìcall and ‚Äìannotate.</p>
</div></blockquote>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">--rrnas</span></code>: PATH &lt;path/to/rRNA.tsv&gt; (See example for format.)</p>
<blockquote>
<div><p>rRNA information will be included in distill output.</p>
</div></blockquote>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">--trnas</span></code>: PATH &lt;path/to/tRNA.tsv&gt; (See example for format.)</p>
<blockquote>
<div><p>tRNA information will be included in distill output.</p>
</div></blockquote>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">--bin_quality</span></code>: PATH &lt;path/to/bin-quality.tsv&gt; (See example for format.)</p>
<blockquote>
<div><p>CheckM and CheckM2 compatible.</p>
</div></blockquote>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">--taxa</span></code>: PATH &lt;path/to/bin-taxonomy.tsv&gt;</p>
<blockquote>
<div><p>Compatible with GTDB. (See example for format.)</p>
</div></blockquote>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">--distill_topic</span></code>: STRING &lt;carbon|energy|misc|nitrogen|transport&gt; OR &lt;default = carbon, energy, misc, nitrogen, transport&gt;</p>
<blockquote>
<div><p>If more than one topic included, they must be enclosed in single quotes.</p>
</div></blockquote>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">--distill_ecosystem</span></code>: STRING &lt;eng_sys|ag&gt;</p>
<blockquote>
<div><p>If more than one ecosystem included, they must be enclosed in single quotes.</p>
</div></blockquote>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">--distill_custom</span></code>: STRING &lt;path/to/custom_distillate.tsv&gt; (See example for format and options.)</p>
<blockquote>
<div><p>As of now, only one custom distillate may be included.</p>
</div></blockquote>
</li>
</ul>
</section>
</section>
<hr class="docutils" />
<section id="select-a-profile-which-suits-your-compute-setup">
<h2>Select a Profile which suits your compute setup<a class="headerlink" href="#select-a-profile-which-suits-your-compute-setup" title="Link to this heading">ÔÉÅ</a></h2>
<p>DRAM2 utilizes either Conda or Singularity for dependency management and the user MUST choose one of the following options on execution of any DRAM2 command</p>
<p><em>The Nextflow profile option is used (`-profile`) - yes! a single hyphen!</em></p>
<ol class="arabic">
<li><div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>-profile<span class="w"> </span>conda
</pre></div>
</div>
<p>This option relies on the local systems Conda. Nextflow will create its own Conda environments to run in.</p>
</li>
<li><div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>-profile<span class="w"> </span>conda_slurm
</pre></div>
</div>
<p>This option will submit each individual DRAM2 process as its own SLURM job. (See Wiki Resource Management for details).
This option relies on the local systems Conda. Nextflow will create its own Conda environments to run in.</p>
</li>
<li><div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>-profile<span class="w"> </span>singularity
</pre></div>
</div>
<p>This option relies on the local systems Singularity. Nextflow will create its own Conda environments to run in.</p>
</li>
<li><div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>-profile<span class="w"> </span>singularity_slurm
</pre></div>
</div>
<p>This option will submit each individual DRAM2 process as its own SLURM job.
This option relies on the local systems Singularity to run the downloaded Singularity container.</p>
</li>
</ol>
<section id="which-is-better">
<h3>Which is Better?<a class="headerlink" href="#which-is-better" title="Link to this heading">ÔÉÅ</a></h3>
<section id="conda-environments">
<h4>Conda Environments<a class="headerlink" href="#conda-environments" title="Link to this heading">ÔÉÅ</a></h4>
<p><strong>Pros:</strong></p>
<ul class="simple">
<li><p>Beginner-Friendly: Easy to install and use, making it accessible for newcomers.</p></li>
<li><p>Reproducibility: Efficient management of environments facilitates reproducibility.</p></li>
</ul>
<p><strong>Cons:</strong></p>
<ul class="simple">
<li><p>Generally slower than using Singularity containers (will have metrics in the future).</p></li>
<li><p>Dependency Conflicts: Dependency resolution can be slow and may lead to conflicts.</p></li>
<li><p>Limited Portability: System dependencies may introduce variability, affecting portability.</p></li>
<li><p>System Variability: Reliance on the host system‚Äôs architecture and libraries can cause variability between systems.</p></li>
</ul>
</section>
<section id="singularity-containers">
<h4>Singularity Containers<a class="headerlink" href="#singularity-containers" title="Link to this heading">ÔÉÅ</a></h4>
<p><strong>Pros:</strong></p>
<ul class="simple">
<li><p>Generally faster than using Conda environments (will have metrics in the future).</p></li>
<li><p>Consistent Environments: Ensures consistent runtime environments, enhancing reproducibility.</p></li>
<li><p>HPC Ideal: Perfect for high-performance computing (HPC) environments without the need for root access.</p></li>
<li><p>Isolation: Offers isolation from the host system, minimizing conflicts.</p></li>
<li><p>Wide Portability: Containers are portable across any Linux system with Singularity.</p></li>
</ul>
<p><strong>Cons:</strong></p>
<ul class="simple">
<li><p>Installation Complexity: Can be trickier to install compared to Conda environments.</p></li>
<li><p>Storage Space: May consume more storage space.</p></li>
</ul>
<p><strong>Summary</strong></p>
<p>Conda is recommended for its ease of use and versatility across different programming languages.
Singularity excels in ensuring reproducibility and compatibility in high-performance computing environments.</p>
</section>
</section>
</section>
<hr class="docutils" />
<section id="dram2-databases">
<h2>DRAM2 Databases<a class="headerlink" href="#dram2-databases" title="Link to this heading">ÔÉÅ</a></h2>
<p>Unlike DRAM1 databases, DRAM2 databases will be pre-formatted and hosted online. Users of DRAM2 will need to:</p>
<ol class="arabic simple">
<li><p>Decide which databases suit their needs.</p></li>
<li><p>Download DRAM2 databases via the provided <code class="docutils literal notranslate"><span class="pre">pull_databases_*.py</span></code> scripts.</p></li>
</ol>
<p><strong>These databases can be quite large, so it‚Äôs important to review the options below.</strong></p>
<p><em>These databases rely on an SQL database of database descriptions, provided in three different sizes based on the user‚Äôs needs.</em></p>
<section id="all-databases-dram2-accommodates">
<h3>All databases DRAM2 accommodates:<a class="headerlink" href="#all-databases-dram2-accommodates" title="Link to this heading">ÔÉÅ</a></h3>
<p><a class="reference external" href="https://www.genome.jp/kegg/">KEGG</a></p>
<ul class="simple">
<li><p>Kyoto Encyclopedia of Genes and Genomes.</p></li>
<li><p>(140G)</p></li>
</ul>
<p><a class="reference external" href="http://bcb.unl.edu/dbCAN2/">dbCAN</a></p>
<ul class="simple">
<li><p>A database for automated carbohydrate-active enzyme annotation.</p></li>
<li><p>(202M)</p></li>
</ul>
<p><a class="reference external" href="https://www.genome.jp/tools/kofamkoala/">Kofam</a></p>
<ul class="simple">
<li><p>Customized HMM database of KEGG Orthologs (KOs).</p></li>
<li><p>(14G)</p></li>
</ul>
<p><a class="reference external" href="https://www.ebi.ac.uk/merops/">MEROPS</a></p>
<ul class="simple">
<li><p>A database of proteolytic enzymes and their substrates.</p></li>
<li><p>(3.6G)</p></li>
</ul>
<p><a class="reference external" href="https://www.ncbi.nlm.nih.gov/genomes/GenomesGroup.cgi?taxid=10239">Viral</a>
- RefSeq viral database.
- (1.6G)</p>
<p><a class="reference external" href="https://github.com/WrightonLabCSU/CAMPER">CAMPER</a></p>
<ul class="simple">
<li><p>Curated Annotations for Microbial (Poly)phenol Enzymes and Reactions.</p></li>
<li><p>(846M)</p></li>
</ul>
<p><a class="reference external" href="https://github.com/dgittins/CANT-HYD-HydrocarbonBiodegradation">CANT-HYD</a></p>
<ul class="simple">
<li><p>Curated database of phylogeny-derived hidden markov models for annotation of marker genes involved in hydrocarbon degradation.</p></li>
<li><p>(877M)</p></li>
</ul>
<p><a class="reference external" href="https://github.com/Arkadiy-Garber/FeGenie">FeGenie</a></p>
<ul class="simple">
<li><p>HMM-based identification and categorization of iron genes and iron gene operons in genomes and metagenome assemblies.</p></li>
<li><p>(6.6M)</p></li>
</ul>
<p><a class="reference external" href="url_to_sulfur_placeholder">Sulfur</a></p>
<ul class="simple">
<li><p>Custom Sulfur database.</p></li>
<li><p>(1.7M)</p></li>
</ul>
<p><a class="reference external" href="url_to_methyl_placeholder">Methyl</a></p>
<ul class="simple">
<li><p>Hidden Markov models (HMMs) based on genes related to iron acquisition, storage, and reduction/oxidation in Bacteria and Archaea.</p></li>
<li><p>(52K)</p></li>
</ul>
<p><a class="reference external" href="https://www.uniprot.org/help/uniref">UniRef</a></p>
<ul class="simple">
<li><p>A comprehensive and non-redundant database of protein sequences.</p></li>
<li><p>(477G)</p></li>
</ul>
<p><a class="reference external" href="https://pfam.xfam.org/">Pfam</a></p>
<ul class="simple">
<li><p>A large collection of protein families, each represented by multiple sequence alignments and hidden Markov models (HMMs).</p></li>
<li><p>(8.8G)</p></li>
</ul>
<p><a class="reference external" href="url_to_vogdb_placeholder">VOGDB</a></p>
<ul class="simple">
<li><p>Placeholder description.</p></li>
<li><p>(4.5G)</p></li>
</ul>
<section id="downloadable-dram2-database-sets">
<h4>Downloadable DRAM2 Database Sets<a class="headerlink" href="#downloadable-dram2-database-sets" title="Link to this heading">ÔÉÅ</a></h4>
<p><strong>Big Set</strong></p>
<ul>
<li><p>Includes all medium databases + UniRef</p></li>
<li><p>Excludes KEGG*</p>
<blockquote>
<div><div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./pull_databases_full.py
</pre></div>
</div>
</div></blockquote>
</li>
</ul>
<p>OR</p>
<p>Follow these instructions to pull manually via <a class="reference external" href="https://www.globus.org/">GLOBUS</a>.</p>
<p><strong>Routine Set</strong></p>
<ul>
<li><p>Includes: dbCAN, Kofam, MEROPS, Viral, CAMPER, CANT-HYD, FeGenie, Sulfur, Methyl, Pfam, VOGDB</p></li>
<li><p>Excludes KEGG*</p></li>
<li><p>Excludes UniRef</p>
<blockquote>
<div><div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./pull_databases_routine.py
</pre></div>
</div>
</div></blockquote>
</li>
</ul>
<p>OR</p>
<p>Follow these instructions to pull manually via <a class="reference external" href="https://www.globus.org/">GLOBUS</a>.</p>
<p><strong>Minimal Set</strong></p>
<ul>
<li><p>Includes: &lt;TBD&gt;</p>
<blockquote>
<div><div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./pull_databases_minimal.py
</pre></div>
</div>
</div></blockquote>
</li>
</ul>
<p>OR</p>
<p>Follow these instructions to pull manually via <a class="reference external" href="https://www.globus.org/">GLOBUS</a>.</p>
</section>
</section>
</section>
<hr class="docutils" />
<section id="slurm-options">
<h2>SLURM Options<a class="headerlink" href="#slurm-options" title="Link to this heading">ÔÉÅ</a></h2>
<p>DRAM2 can utilize <a class="reference external" href="https://slurm.schedmd.com/documentation.html">SLURM</a> for scheduling jobs.</p>
<p>SLURM is used when either of these <code class="docutils literal notranslate"><span class="pre">--profile</span></code> options are used: <cite>conda_slurm</cite> or <cite>singularity_slurm</cite></p>
<ul class="simple">
<li><p>Each DRAM2 process is submitted as an individual SLURM job</p></li>
<li><p>This means DRAM2 can scale horizontally on an HPC</p></li>
</ul>
<section id="slurm-command-line-options">
<h3>SLURM command-line options<a class="headerlink" href="#slurm-command-line-options" title="Link to this heading">ÔÉÅ</a></h3>
<p>Specify SLURM job duration:</p>
<blockquote>
<div><p><code class="docutils literal notranslate"><span class="pre">--time</span> <span class="pre">&lt;time</span> <span class="pre">in</span> <span class="pre">SLURM</span> <span class="pre">format&gt;</span></code></p>
<p>Example: <code class="docutils literal notranslate"><span class="pre">--time</span> <span class="pre">10h</span></code></p>
<p>Example: <code class="docutils literal notranslate"><span class="pre">--time</span> <span class="pre">7d</span></code></p>
</div></blockquote>
<p>Specify a specific node to compute on:</p>
<blockquote>
<div><p><code class="docutils literal notranslate"><span class="pre">--slurm_node</span> <span class="pre">&lt;node_name&gt;</span></code></p>
<p>Example: <code class="docutils literal notranslate"><span class="pre">--slurm_node</span> <span class="pre">alpha</span></code></p>
</div></blockquote>
<p>Specify a specific partition/queue to compute within:</p>
<blockquote>
<div><p><code class="docutils literal notranslate"><span class="pre">--slurm_queue</span> <span class="pre">&lt;node_name&gt;</span></code></p>
<p>Example: <code class="docutils literal notranslate"><span class="pre">--slurm_queue</span> <span class="pre">smith</span></code></p>
<p>Example: <code class="docutils literal notranslate"><span class="pre">--slurm_queue</span> <span class="pre">'smith-hi,smith-low'</span></code></p>
</div></blockquote>
<p><strong>Note:</strong> SLURM can be tricky because administrators do not set up SLURM the same across machines.
You may need to modify a given profile <cite>.config</cite> file to get DRAM2 with SLURM to work on your HPC.</p>
<p><strong>For Example</strong>, a HPC our group computes on does not allow for memory specificaton. Thus, to use DRAM2 with SLURM on this HPC, we comment out (‚Äú//‚Äù) the <code class="docutils literal notranslate"><span class="pre">memory=</span></code> lines in the config files.</p>
<p>The config files are located in:
<code class="docutils literal notranslate"><span class="pre">/assets/conda/conda_slurm.config</span></code> and <code class="docutils literal notranslate"><span class="pre">/assets/singularity/singularity_slurm.config</span></code></p>
</section>
</section>
<hr class="docutils" />
<section id="software-used">
<h2>Software Used<a class="headerlink" href="#software-used" title="Link to this heading">ÔÉÅ</a></h2>
<ul class="simple">
<li><p><strong>BBTools</strong> <a class="reference external" href="https://jgi.doe.gov/data-and-tools/bbtools/">v39.01</a></p></li>
<li><p><strong>Bowtie2</strong> <a class="reference external" href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">v2.5.1</a></p></li>
<li><p><strong>Prodigal</strong> <a class="reference external" href="https://github.com/hyattpd/Prodigal">v2.6.3</a></p></li>
<li><p><strong>Python</strong> <a class="reference external" href="https://www.python.org/downloads/release/python-3100/">v3.10</a></p></li>
<li><p><strong>Pandas</strong> <a class="reference external" href="https://pandas.pydata.org/pandas-docs/version/1.5.2/">v1.5.2</a></p></li>
<li><p><strong>Pytest</strong> <a class="reference external" href="https://docs.pytest.org/en/7.2.x/">v7.2.0</a></p></li>
<li><p><strong>Scikit-bio</strong> <a class="reference external" href="http://scikit-bio.org/">v0.5.7</a></p></li>
<li><p><strong>MMseqs2</strong> <a class="reference external" href="https://github.com/soedinglab/MMseqs2">v14.7e284</a></p></li>
<li><p><strong>HMMER</strong> <a class="reference external" href="http://hmmer.org/">v3.3.2</a></p></li>
<li><p><strong>SciPy</strong> <a class="reference external" href="https://www.scipy.org/">v1.8.1</a></p></li>
<li><p><strong>SQLAlchemy</strong> <a class="reference external" href="https://www.sqlalchemy.org/">v1.4.46</a></p></li>
<li><p><strong>Barrnap</strong> <a class="reference external" href="https://github.com/tseemann/barrnap">v0.9</a></p></li>
<li><p><strong>Altair</strong> <a class="reference external" href="https://altair-viz.github.io/">v4.2.0</a></p></li>
<li><p><strong>OpenPyXL</strong> <a class="reference external" href="https://openpyxl.readthedocs.io/en/stable/">v3.0.10</a></p></li>
<li><p><strong>NetworkX</strong> <a class="reference external" href="https://networkx.org/">v2.8.8</a></p></li>
<li><p><strong>Ruby</strong> <a class="reference external" href="https://www.ruby-lang.org/en/downloads/">v3.1.2</a></p></li>
<li><p><strong>GNU Parallel</strong> <a class="reference external" href="https://www.gnu.org/software/parallel/">v20221122</a></p></li>
<li><p><strong>tRNAscan-SE</strong> <a class="reference external" href="http://lowelab.ucsc.edu/tRNAscan-SE/">v2.0.12</a></p></li>
<li><p><strong>Samtools</strong> <a class="reference external" href="http://www.htslib.org/">v1.17</a></p></li>
<li><p><strong>CD-HIT</strong> <a class="reference external" href="http://weizhong-lab.ucsd.edu/cd-hit/">v4.6</a></p></li>
<li><p><strong>CoverM</strong> <a class="reference external" href="https://github.com/wwood/CoverM">v0.6.1</a></p></li>
<li><p><strong>Subread</strong> <a class="reference external" href="http://subread.sourceforge.net/">v2.0.6</a></p></li>
<li><p><strong>XlsxWriter</strong> <a class="reference external" href="https://xlsxwriter.readthedocs.io/">v3.1.6</a></p></li>
<li><p><strong>Numpy</strong> <a class="reference external" href="https://numpy.org/">v1.26.0</a></p></li>
</ul>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="example_output.html" class="btn btn-neutral float-left" title="Output Files" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="nextflow_options.html" class="btn btn-neutral float-right" title="Nextflow Pipeline Options" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Rory M. Flynn and Reed R. Woyda.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>